crates/table/src/bflatn_to_bsatn_fast_path.rs

```diff
@@ -0,0 +1,424 @@
+                        AlgebraicType::U8,
+                        AlgebraicType::U8,
+                        AlgebraicType::U8,
+                    ]),
+                    AlgebraicType::product([AlgebraicType::U16, AlgebraicType::U16]),
+                    AlgebraicType::U32,
+                ])]),
+                5,
+                // Sums get wonky layouts
+                // because BFLATN and BSATN store the tag and the payload in opposite orders.
+                &[(4, 0, 1), (0, 1, 4)][..],
+            ),
+            (
+                ProductType::from([
+                    AlgebraicType::sum([AlgebraicType::U128, AlgebraicType::I128]),
+                    AlgebraicType::U32,
+                ]),
+                21,
+                // Sums get wonky layouts
+                // because BFLATN and BSATN store the tag and the payload in opposite orders.
+                &[(16, 0, 1), (0, 1, 16), (32, 17, 4)][..],
+            ),
+            (
+                ProductType::from([
+                    AlgebraicType::U128,
+                    AlgebraicType::U64,
+                    AlgebraicType::U32,
+                    AlgebraicType::U16,
+                    AlgebraicType::U8,
+                ]),
+                31,
+                &[(0, 0, 31)][..],
+            ),
+            (
+                ProductType::from([
+                    AlgebraicType::U8,
+                    AlgebraicType::U16,
+                    AlgebraicType::U32,
+                    AlgebraicType::U64,
+                    AlgebraicType::U128,
+                ]),
+                31,
+                &[(0, 0, 1), (2, 1, 30)][..],
+            ),
+        ] {
+            assert_expected_layout(ty, bsatn_length, fields);
+        }
+    }
+
+    #[test]
+    fn known_types_not_applicable() {
+        for ty in [
+            AlgebraicType::String,
+            AlgebraicType::bytes(),
+            AlgebraicType::never(),
+            AlgebraicType::array(AlgebraicType::U16),
+            AlgebraicType::map(AlgebraicType::U8, AlgebraicType::I8),
+            AlgebraicType::sum([AlgebraicType::U8, AlgebraicType::U16]),
+        ] {
+            let layout = RowTypeLayout::from(ProductType::from([ty]));
+            if let Some(computed) = KnownBsatnLayout::for_row_type(&layout) {
+                panic!("Expected row type not to have a constant BSATN layout!\nRow type: {layout:#?}\nBSATN layout: {computed:#?}");
+            }
+        }
+    }
+
+    proptest! {
+        // The test `known_bsatn_same_as_bflatn_from` generates a lot of rejects,
+        // as a vast majority of the space of `ProductType` does not have a fixed BSATN length.
+        // Writing a proptest generator which produces only types that have a fixed BSATN length
+        // seems hard, because we'd have to generate sums with known matching layouts,
+        // so we just bump the `max_global_rejects` up as high as it'll go and move on with our lives.
```

Comment:
I would prefer this comment to explain why this test is needed over `known_types_expected_layout`. Because from my reading, it doesn't seem like it gives us any further _observable_ test coverage.

Reply:
Comment expanded; now in full reads:
```rust
// The test `known_bsatn_same_as_bflatn_from` generates a lot of rejects,
// as a vast majority of the space of `ProductType` does not have a fixed BSATN length.
// Writing a proptest generator which produces only types that have a fixed BSATN length
// seems hard, because we'd have to generate sums with known matching layouts,
// so we just bump the `max_global_rejects` up as high as it'll go and move on with our lives.
//
// Note that I (pgoldman 2024-03-21) tried modifying `generate_typed_row`
// to not emit `String`, `Array` or `Map` types (the trivially var-len types),
// but did not see a meaningful decrease in the number of rejects.
// This is because a majority of the var-len BSATN types in the `generate_typed_row` space
// are due to sums with inconsistent payload layouts.
//
// We still include the test `known_bsatn_same_as_bsatn_from`
// because it tests row types not covered in `known_types_expected_layout`,
// especially larger types with unusual sequences of aligned fields.
```

Reply:
> // We still include the test `known_bsatn_same_as_bsatn_from`
// because it tests row types not covered in `known_types_expected_layout`,
// especially larger types with unusual sequences of aligned fields.

But are we guaranteed to catch any such failures?

Reply:
This is really more of an existential question about how much you trust statistical testing. I believe this test will catch any such failures at a high rate.